File: wrappers/php/foundation/src/RecipientCipher.php
<?php
/**
* Copyright (C) 2015-2019 Virgil Security, Inc.
*
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions are
* met:
*
* (1) Redistributions of source code must retain the above copyright
* notice, this list of conditions and the following disclaimer.
*
* (2) Redistributions in binary form must reproduce the above copyright
* notice, this list of conditions and the following disclaimer in
* the documentation and/or other materials provided with the
* distribution.
*
* (3) Neither the name of the copyright holder nor the names of its
* contributors may be used to endorse or promote products derived from
* this software without specific prior written permission.
*
* THIS SOFTWARE IS PROVIDED BY THE AUTHOR ''AS IS'' AND ANY EXPRESS OR
* IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
* WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
* DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
* INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
* (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
* HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
* STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
* IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
* POSSIBILITY OF SUCH DAMAGE.
*
* Lead Maintainer: Virgil Security Inc. <support@virgilsecurity.com>
*/

namespace VirgilCrypto\Foundation;

/**
* This class provides hybrid encryption algorithm that combines symmetric
* cipher for data encryption and asymmetric cipher and password based
* cipher for symmetric key encryption.
*/
class RecipientCipher
{

    /**
    * @var
    */
    private $ctx;

    /**
    * Create underlying C context.
    * @param null $ctx
    * @return void
    */
    public function __construct($ctx = null)
    {
        $this->ctx = is_null($ctx) ? vscf_recipient_cipher_new_php() : $ctx;
    }

    /**
    * Destroy underlying C context.
    * @return void
    */
    public function __destructor()
    {
        vscf_recipient_cipher_delete_php($this->ctx);
    }

    /**
    * @param Random $random
    * @return void
    */
    public function useRandom(Random $random): void
    {
        vscf_recipient_cipher_use_random_php($this->ctx, $random->getCtx());
    }

    /**
    * @param Cipher $encryptionCipher
    * @return void
    */
    public function useEncryptionCipher(Cipher $encryptionCipher): void
    {
        vscf_recipient_cipher_use_encryption_cipher_php($this->ctx, $encryptionCipher->getCtx());
    }

    /**
    * @param Hash $signerHash
    * @return void
    */
    public function useSignerHash(Hash $signerHash): void
    {
        vscf_recipient_cipher_use_signer_hash_php($this->ctx, $signerHash->getCtx());
    }

    /**
    * Return true if a key recipient with a given id has been added.
    * Note, operation has O(N) time complexity.
    *
    * @param string $recipientId
    * @return bool
    */
    public function hasKeyRecipient(string $recipientId): bool
    {
        return vscf_recipient_cipher_has_key_recipient_php($this->ctx, $recipientId);
    }

    /**
    * Add recipient defined with id and public key.
    *
    * @param string $recipientId
    * @param PublicKey $publicKey
    * @return void
    */
    public function addKeyRecipient(string $recipientId, PublicKey $publicKey): void
    {
        vscf_recipient_cipher_add_key_recipient_php($this->ctx, $recipientId, $publicKey->getCtx());
    }

    /**
    * Remove all recipients.
    *
    * @return void
    */
    public function clearRecipients(): void
    {
        vscf_recipient_cipher_clear_recipients_php($this->ctx);
    }

    /**
    * Add identifier and private key to sign initial plain text.
    * Return error if the private key can not sign.
    *
    * @param string $signerId
    * @param PrivateKey $privateKey
    * @return void
    * @throws \Exception
    */
    public function addSigner(string $signerId, PrivateKey $privateKey): void
    {
        vscf_recipient_cipher_add_signer_php($this->ctx, $signerId, $privateKey->getCtx());
    }

    /**
    * Remove all signers.
    *
    * @return void
    */
    public function clearSigners(): void
    {
        vscf_recipient_cipher_clear_signers_php($this->ctx);
    }

    /**
    * Provide access to the custom params object.
    * The returned object can be used to add custom params or read it.
    *
    * @return MessageInfoCustomParams
    */
    public function customParams(): MessageInfoCustomParams
    {
        $ctx = vscf_recipient_cipher_custom_params_php($this->ctx);
        return new MessageInfoCustomParams($ctx);
    }

    /**
    * Start encryption process.
    *
    * @return void
    * @throws \Exception
    */
    public function startEncryption(): void
    {
        vscf_recipient_cipher_start_encryption_php($this->ctx);
    }

    /**
    * Start encryption process with known plain text size.
    *
    * Precondition: At least one signer should be added.
    * Note, store message info footer as well.
    *
    * @param int $dataSize
    * @return void
    * @throws \Exception
    */
    public function startSignedEncryption(int $dataSize): void
    {
        vscf_recipient_cipher_start_signed_encryption_php($this->ctx, $dataSize);
    }

    /**
    * Return buffer length required to hold message info returned by the
    * "pack message info" method.
    * Precondition: all recipients and custom parameters should be set.
    *
    * @return int
    */
    public function messageInfoLen(): int
    {
        return vscf_recipient_cipher_message_info_len_php($this->ctx);
    }

    /**
    * Return serialized message info to the buffer.
    *
    * Precondition: this method should be called after "start encryption".
    * Precondition: this method should be called before "finish encryption".
    *
    * Note, store message info to use it for decryption process,
    * or place it at the encrypted data beginning (embedding).
    *
    * Return message info - recipients public information,
    * algorithm information, etc.
    *
    * @return string
    */
    public function packMessageInfo(): string
    {
        return vscf_recipient_cipher_pack_message_info_php($this->ctx);
    }

    /**
    * Return buffer length required to hold output of the method
    * "process encryption" and method "finish" during encryption.
    *
    * @param int $dataLen
    * @return int
    */
    public function encryptionOutLen(int $dataLen): int
    {
        return vscf_recipient_cipher_encryption_out_len_php($this->ctx, $dataLen);
    }

    /**
    * Process encryption of a new portion of data.
    *
    * @param string $data
    * @return string
    * @throws \Exception
    */
    public function processEncryption(string $data): string
    {
        return vscf_recipient_cipher_process_encryption_php($this->ctx, $data);
    }

    /**
    * Accomplish encryption.
    *
    * @return string
    * @throws \Exception
    */
    public function finishEncryption(): string
    {
        return vscf_recipient_cipher_finish_encryption_php($this->ctx);
    }

    /**
    * Initiate decryption process with a recipient private key.
    * Message Info can be empty if it was embedded to encrypted data.
    *
    * @param string $recipientId
    * @param PrivateKey $privateKey
    * @param string $messageInfo
    * @return void
    * @throws \Exception
    */
    public function startDecryptionWithKey(string $recipientId, PrivateKey $privateKey, string $messageInfo): void
    {
        vscf_recipient_cipher_start_decryption_with_key_php($this->ctx, $recipientId, $privateKey->getCtx(), $messageInfo);
    }

    /**
    * Initiate decryption process with a recipient private key.
    * Message Info can be empty if it was embedded to encrypted data.
    * Message Info footer can be empty if it was embedded to encrypted data.
    * If footer was embedded, method "start decryption with key" can be used.
    *
    * @param string $recipientId
    * @param PrivateKey $privateKey
    * @param string $messageInfo
    * @param string $messageInfoFooter
    * @return void
    * @throws \Exception
    */
    public function startVerifiedDecryptionWithKey(string $recipientId, PrivateKey $privateKey, string $messageInfo, string $messageInfoFooter): void
    {
        vscf_recipient_cipher_start_verified_decryption_with_key_php($this->ctx, $recipientId, $privateKey->getCtx(), $messageInfo, $messageInfoFooter);
    }

    /**
    * Return buffer length required to hold output of the method
    * "process decryption" and method "finish" during decryption.
    *
    * @param int $dataLen
    * @return int
    */
    public function decryptionOutLen(int $dataLen): int
    {
        return vscf_recipient_cipher_decryption_out_len_php($this->ctx, $dataLen);
    }

    /**
    * Process with a new portion of data.
    * Return error if data can not be encrypted or decrypted.
    *
    * @param string $data
    * @return string
    * @throws \Exception
    */
    public function processDecryption(string $data): string
    {
        return vscf_recipient_cipher_process_decryption_php($this->ctx, $data);
    }

    /**
    * Accomplish decryption.
    *
    * @return string
    * @throws \Exception
    */
    public function finishDecryption(): string
    {
        return vscf_recipient_cipher_finish_decryption_php($this->ctx);
    }

    /**
    * Return true if data was signed by a sender.
    *
    * Precondition: this method should be called after "finish decryption".
    *
    * @return bool
    */
    public function isDataSigned(): bool
    {
        return vscf_recipient_cipher_is_data_signed_php($this->ctx);
    }

    /**
    * Return information about signers that sign data.
    *
    * Precondition: this method should be called after "finish decryption".
    * Precondition: method "is data signed" returns true.
    *
    * @return SignerInfoList
    */
    public function signerInfos(): SignerInfoList
    {
        $ctx = vscf_recipient_cipher_signer_infos_php($this->ctx);
        return new SignerInfoList($ctx);
    }

    /**
    * Verify given cipher info.
    *
    * @param SignerInfo $signerInfo
    * @param PublicKey $publicKey
    * @return bool
    */
    public function verifySignerInfo(SignerInfo $signerInfo, PublicKey $publicKey): bool
    {
        return vscf_recipient_cipher_verify_signer_info_php($this->ctx, $signerInfo->getCtx(), $publicKey->getCtx());
    }

    /**
    * Return buffer length required to hold message footer returned by the
    * "pack message footer" method.
    *
    * Precondition: this method should be called after "finish encryption".
    *
    * @return int
    */
    public function messageInfoFooterLen(): int
    {
        return vscf_recipient_cipher_message_info_footer_len_php($this->ctx);
    }

    /**
    * Return serialized message info footer to the buffer.
    *
    * Precondition: this method should be called after "finish encryption".
    *
    * Note, store message info to use it for verified decryption process,
    * or place it at the encrypted data ending (embedding).
    *
    * Return message info footer - signers public information, etc.
    *
    * @return string
    * @throws \Exception
    */
    public function packMessageInfoFooter(): string
    {
        return vscf_recipient_cipher_pack_message_info_footer_php($this->ctx);
    }

    /**
    * Get C context.
    *
    * @return resource
    */
    public function getCtx()
    {
        return $this->ctx;
    }
}

File: wrappers/cpp/src/foundation/recipient_cipher.cpp
// Copyright (C) 2015-2026 Virgil Security, Inc.
//
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
//     (1) Redistributions of source code must retain the above copyright
//     notice, this list of conditions and the following disclaimer.
//
//     (2) Redistributions in binary form must reproduce the above copyright
//     notice, this list of conditions and the following disclaimer in
//     the documentation and/or other materials provided with the
//     distribution.
//
//     (3) Neither the name of the copyright holder nor the names of its
//     contributors may be used to endorse or promote products derived from
//     this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE AUTHOR ''AS IS'' AND ANY EXPRESS OR
// IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
// DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
// INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
// IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.
//
// Lead Maintainer: Virgil Security Inc. <support@virgilsecurity.com>

#include <virgil/crypto/foundation/recipient_cipher.hpp>
#include <virgil/crypto/foundation/vscf_recipient_cipher.h>
#include <virgil/crypto/foundation/vscf_impl.h>
#include <virgil/crypto/foundation/cipher.hpp>
#include <virgil/crypto/foundation/hash.hpp>
#include <virgil/crypto/foundation/key_wrap.hpp>
#include <virgil/crypto/foundation/message_info_custom_params.hpp>
#include <virgil/crypto/foundation/padding.hpp>
#include <virgil/crypto/foundation/padding_params.hpp>
#include <virgil/crypto/foundation/private_key.hpp>
#include <virgil/crypto/foundation/public_key.hpp>
#include <virgil/crypto/foundation/random.hpp>
#include <virgil/crypto/foundation/signer_info.hpp>
#include <virgil/crypto/foundation/signer_info_list.hpp>
#include <virgil/crypto/common/vsc_buffer.h>

namespace virgil::crypto::foundation {

RecipientCipher::RecipientCipher() : c_ctx_(vscf_recipient_cipher_new()) {}

RecipientCipher::RecipientCipher(vscf_recipient_cipher_t* c_ctx) noexcept : c_ctx_(c_ctx) {}

RecipientCipher::RecipientCipher(const RecipientCipher& other) : c_ctx_(vscf_recipient_cipher_shallow_copy(other.c_ctx_)) {}

RecipientCipher::RecipientCipher(RecipientCipher&& other) noexcept : c_ctx_(other.c_ctx_) { other.c_ctx_ = nullptr; }

RecipientCipher& RecipientCipher::operator=(const RecipientCipher& other) {
    if (this != &other) {
        vscf_recipient_cipher_delete(c_ctx_);
        c_ctx_ = vscf_recipient_cipher_shallow_copy(other.c_ctx_);
    }
    return *this;
}

RecipientCipher& RecipientCipher::operator=(RecipientCipher&& other) noexcept {
    if (this != &other) {
        vscf_recipient_cipher_delete(c_ctx_);
        c_ctx_ = other.c_ctx_;
        other.c_ctx_ = nullptr;
    }
    return *this;
}

RecipientCipher::~RecipientCipher() { vscf_recipient_cipher_delete(c_ctx_); }

vscf_recipient_cipher_t* RecipientCipher::c_ctx() const noexcept { return c_ctx_; }

void RecipientCipher::set_random(const Random& random) {
    vscf_recipient_cipher_release_random(c_ctx_);
    vscf_recipient_cipher_use_random(c_ctx_, random.impl());
}

void RecipientCipher::set_encryption_cipher(const Cipher& encryption_cipher) {
    vscf_recipient_cipher_release_encryption_cipher(c_ctx_);
    vscf_recipient_cipher_use_encryption_cipher(c_ctx_, encryption_cipher.impl());
}

void RecipientCipher::set_encryption_padding(const Padding& encryption_padding) {
    vscf_recipient_cipher_release_encryption_padding(c_ctx_);
    vscf_recipient_cipher_use_encryption_padding(c_ctx_, encryption_padding.impl());
}

void RecipientCipher::set_padding_params(const PaddingParams& padding_params) {
    vscf_recipient_cipher_release_padding_params(c_ctx_);
    vscf_recipient_cipher_use_padding_params(c_ctx_, padding_params.c_ctx());
}

void RecipientCipher::set_signer_hash(const Hash& signer_hash) {
    vscf_recipient_cipher_release_signer_hash(c_ctx_);
    vscf_recipient_cipher_use_signer_hash(c_ctx_, signer_hash.impl());
}

bool RecipientCipher::has_key_recipient(std::span<const uint8_t> recipient_id) const {
    auto proxy_result = vscf_recipient_cipher_has_key_recipient(c_ctx_, recipient_id.empty() ? vsc_data_empty() : vsc_data(recipient_id.data(), recipient_id.size()));
    return proxy_result;
}

void RecipientCipher::add_key_recipient(std::span<const uint8_t> recipient_id, const PublicKey& public_key) {
    vscf_recipient_cipher_add_key_recipient(c_ctx_, recipient_id.empty() ? vsc_data_empty() : vsc_data(recipient_id.data(), recipient_id.size()), public_key.impl());
}

void RecipientCipher::add_kek_recipient(std::span<const uint8_t> kek_id, std::span<const uint8_t> kek, const KeyWrap& key_wrap) {
    vscf_recipient_cipher_add_kek_recipient(c_ctx_, kek_id.empty() ? vsc_data_empty() : vsc_data(kek_id.data(), kek_id.size()), kek.empty() ? vsc_data_empty() : vsc_data(kek.data(), kek.size()), key_wrap.impl());
}

void RecipientCipher::clear_recipients() {
    vscf_recipient_cipher_clear_recipients(c_ctx_);
}

tl::expected<void, Error> RecipientCipher::add_signer(std::span<const uint8_t> signer_id, const PrivateKey& private_key) {
    const vscf_status_t status = vscf_recipient_cipher_add_signer(c_ctx_, signer_id.empty() ? vsc_data_empty() : vsc_data(signer_id.data(), signer_id.size()), private_key.impl());
    if (status != vscf_status_SUCCESS) {
        return tl::unexpected(static_cast<Error>(status));
    }
    return {};
}

void RecipientCipher::clear_signers() {
    vscf_recipient_cipher_clear_signers(c_ctx_);
}

MessageInfoCustomParams RecipientCipher::custom_params() {
    auto proxy_result = vscf_recipient_cipher_custom_params(c_ctx_);
    return MessageInfoCustomParams(vscf_message_info_custom_params_shallow_copy(const_cast<vscf_message_info_custom_params_t*>(proxy_result)));
}

tl::expected<void, Error> RecipientCipher::start_encryption() {
    const vscf_status_t status = vscf_recipient_cipher_start_encryption(c_ctx_);
    if (status != vscf_status_SUCCESS) {
        return tl::unexpected(static_cast<Error>(status));
    }
    return {};
}

tl::expected<void, Error> RecipientCipher::start_signed_encryption(std::size_t data_size) {
    const vscf_status_t status = vscf_recipient_cipher_start_signed_encryption(c_ctx_, data_size);
    if (status != vscf_status_SUCCESS) {
        return tl::unexpected(static_cast<Error>(status));
    }
    return {};
}

std::size_t RecipientCipher::message_info_len() const {
    auto proxy_result = vscf_recipient_cipher_message_info_len(c_ctx_);
    return proxy_result;
}

std::vector<uint8_t> RecipientCipher::pack_message_info() {
    std::vector<uint8_t> message_info(this->message_info_len());
    vsc_buffer_t* message_info_buf = vsc_buffer_new();
    vsc_buffer_use(message_info_buf, message_info.data(), message_info.size());
    vscf_recipient_cipher_pack_message_info(c_ctx_, message_info_buf);
    message_info.resize(vsc_buffer_len(message_info_buf));
    vsc_buffer_delete(message_info_buf);
    return message_info;
}

std::size_t RecipientCipher::encryption_out_len(std::size_t data_len) {
    auto proxy_result = vscf_recipient_cipher_encryption_out_len(c_ctx_, data_len);
    return proxy_result;
}

tl::expected<std::vector<uint8_t>, Error> RecipientCipher::process_encryption(std::span<const uint8_t> data) {
    std::vector<uint8_t> out(this->encryption_out_len(data.size()));
    vsc_buffer_t* out_buf = vsc_buffer_new();
    vsc_buffer_use(out_buf, out.data(), out.size());
    const vscf_status_t status = vscf_recipient_cipher_process_encryption(c_ctx_, data.empty() ? vsc_data_empty() : vsc_data(data.data(), data.size()), out_buf);
    out.resize(vsc_buffer_len(out_buf));
    vsc_buffer_delete(out_buf);
    if (status != vscf_status_SUCCESS) {
        return tl::unexpected(static_cast<Error>(status));
    }
    return out;
}

tl::expected<std::vector<uint8_t>, Error> RecipientCipher::finish_encryption() {
    std::vector<uint8_t> out(this->encryption_out_len(0));
    vsc_buffer_t* out_buf = vsc_buffer_new();
    vsc_buffer_use(out_buf, out.data(), out.size());
    const vscf_status_t status = vscf_recipient_cipher_finish_encryption(c_ctx_, out_buf);
    out.resize(vsc_buffer_len(out_buf));
    vsc_buffer_delete(out_buf);
    if (status != vscf_status_SUCCESS) {
        return tl::unexpected(static_cast<Error>(status));
    }
    return out;
}

tl::expected<void, Error> RecipientCipher::start_decryption_with_kek(std::span<const uint8_t> kek_id, std::span<const uint8_t> kek, const KeyWrap& key_wrap, std::span<const uint8_t> message_info) {
    const vscf_status_t status = vscf_recipient_cipher_start_decryption_with_kek(c_ctx_, kek_id.empty() ? vsc_data_empty() : vsc_data(kek_id.data(), kek_id.size()), kek.empty() ? vsc_data_empty() : vsc_data(kek.data(), kek.size()), key_wrap.impl(), message_info.empty() ? vsc_data_empty() : vsc_data(message_info.data(), message_info.size()));
    if (status != vscf_status_SUCCESS) {
        return tl::unexpected(static_cast<Error>(status));
    }
    return {};
}

tl::expected<void, Error> RecipientCipher::start_decryption_with_key(std::span<const uint8_t> recipient_id, const PrivateKey& private_key, std::span<const uint8_t> message_info) {
    const vscf_status_t status = vscf_recipient_cipher_start_decryption_with_key(c_ctx_, recipient_id.empty() ? vsc_data_empty() : vsc_data(recipient_id.data(), recipient_id.size()), private_key.impl(), message_info.empty() ? vsc_data_empty() : vsc_data(message_info.data(), message_info.size()));
    if (status != vscf_status_SUCCESS) {
        return tl::unexpected(static_cast<Error>(status));
    }
    return {};
}

tl::expected<void, Error> RecipientCipher::start_verified_decryption_with_key(std::span<const uint8_t> recipient_id, const PrivateKey& private_key, std::span<const uint8_t> message_info, std::span<const uint8_t> message_info_footer) {
    const vscf_status_t status = vscf_recipient_cipher_start_verified_decryption_with_key(c_ctx_, recipient_id.empty() ? vsc_data_empty() : vsc_data(recipient_id.data(), recipient_id.size()), private_key.impl(), message_info.empty() ? vsc_data_empty() : vsc_data(message_info.data(), message_info.size()), message_info_footer.empty() ? vsc_data_empty() : vsc_data(message_info_footer.data(), message_info_footer.size()));
    if (status != vscf_status_SUCCESS) {
        return tl::unexpected(static_cast<Error>(status));
    }
    return {};
}

std::size_t RecipientCipher::decryption_out_len(std::size_t data_len) {
    auto proxy_result = vscf_recipient_cipher_decryption_out_len(c_ctx_, data_len);
    return proxy_result;
}

tl::expected<std::vector<uint8_t>, Error> RecipientCipher::process_decryption(std::span<const uint8_t> data) {
    std::vector<uint8_t> out(this->decryption_out_len(data.size()));
    vsc_buffer_t* out_buf = vsc_buffer_new();
    vsc_buffer_use(out_buf, out.data(), out.size());
    const vscf_status_t status = vscf_recipient_cipher_process_decryption(c_ctx_, data.empty() ? vsc_data_empty() : vsc_data(data.data(), data.size()), out_buf);
    out.resize(vsc_buffer_len(out_buf));
    vsc_buffer_delete(out_buf);
    if (status != vscf_status_SUCCESS) {
        return tl::unexpected(static_cast<Error>(status));
    }
    return out;
}

tl::expected<std::vector<uint8_t>, Error> RecipientCipher::finish_decryption() {
    std::vector<uint8_t> out(this->decryption_out_len(0));
    vsc_buffer_t* out_buf = vsc_buffer_new();
    vsc_buffer_use(out_buf, out.data(), out.size());
    const vscf_status_t status = vscf_recipient_cipher_finish_decryption(c_ctx_, out_buf);
    out.resize(vsc_buffer_len(out_buf));
    vsc_buffer_delete(out_buf);
    if (status != vscf_status_SUCCESS) {
        return tl::unexpected(static_cast<Error>(status));
    }
    return out;
}

bool RecipientCipher::is_data_signed() const {
    auto proxy_result = vscf_recipient_cipher_is_data_signed(c_ctx_);
    return proxy_result;
}

SignerInfoList RecipientCipher::signer_infos() const {
    auto proxy_result = vscf_recipient_cipher_signer_infos(c_ctx_);
    return SignerInfoList(vscf_signer_info_list_shallow_copy(const_cast<vscf_signer_info_list_t*>(proxy_result)));
}

bool RecipientCipher::verify_signer_info(const SignerInfo& signer_info, const PublicKey& public_key) {
    auto proxy_result = vscf_recipient_cipher_verify_signer_info(c_ctx_, signer_info.c_ctx(), public_key.impl());
    return proxy_result;
}

std::size_t RecipientCipher::message_info_footer_len() const {
    auto proxy_result = vscf_recipient_cipher_message_info_footer_len(c_ctx_);
    return proxy_result;
}

tl::expected<std::vector<uint8_t>, Error> RecipientCipher::pack_message_info_footer() {
    std::vector<uint8_t> out(this->message_info_footer_len());
    vsc_buffer_t* out_buf = vsc_buffer_new();
    vsc_buffer_use(out_buf, out.data(), out.size());
    const vscf_status_t status = vscf_recipient_cipher_pack_message_info_footer(c_ctx_, out_buf);
    out.resize(vsc_buffer_len(out_buf));
    vsc_buffer_delete(out_buf);
    if (status != vscf_status_SUCCESS) {
        return tl::unexpected(static_cast<Error>(status));
    }
    return out;
}

}  // namespace virgil::crypto::foundation
